File: src/main/java/net/xalcon/technomage/client/gui/technonomicon/GuiTechnonomiconResearchViewer.java
package net.xalcon.technomage.client.gui.technonomicon;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.init.Blocks;
import net.minecraft.init.Items;
import net.minecraft.util.ResourceLocation;
import net.xalcon.technomage.Technomage;
import net.xalcon.technomage.api.gui.technonomicon.*;
import org.lwjgl.opengl.GL11;
import org.lwjgl.util.Rectangle;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class GuiTechnonomiconResearchViewer extends GuiScreen
{
    private final static ResourceLocation RESEARCH_TEX = new ResourceLocation(Technomage.MOD_ID, "textures/gui/technonomicon/research.png");
    private final static int INDENT = 8;
    private List<IResearchView> viewList = new ArrayList<>();
    private GuiTechnonomiconTabList tabList = new GuiTechnonomiconTabList();

    private int winScaleFactor;
    private int guiWidth;
    private int guiHeight;
    private int guiLeft;
    private int guiTop;

    @Override
    public void initGui()
    {
        super.initGui();
        this.guiWidth = 256 + 96;
        this.guiHeight = 224;

        this.guiLeft = (this.width - this.guiWidth) / 2;
        this.guiTop = (this.height - this.guiHeight) / 2;

        this.viewList.clear();

        ResearchView view = new ResearchView(new ResourceLocation(Technomage.MOD_ID, "test"), new ItemStackGuiIcon(Items.GOLD_NUGGET));
        view.addResearchItem(new IResearchViewItem()
        {
            @Override
            public Rectangle getBounds()
            {
                return new Rectangle(256, 256, 16, 16);
            }

            @Override
            public TechnonomiconTooltip getTooltip()
            {
                return new TechnonomiconTooltip("ITEM!", "Yep, just a test.");
            }

            @Override
            public ResearchState getResearchState()
            {
                return ResearchState.NOT_RESEARCHED;
            }

            @Override
            public ResourceLocation getRegistryName()
            {
                return new ResourceLocation(Technomage.MOD_ID, "testitem");
            }

            @Override
            public IGuiIcon getIcon()
            {
                return new ItemStackGuiIcon(Items.GOLDEN_PICKAXE);
            }

            @Override
            public ResearchIconType getResearchIconType()
            {
                return ResearchIconType.DEFAULT;
            }

            @Override
            public void onClicked()
            {
            }
        });
        this.viewList.add(view);

        view = new ResearchView(new ResourceLocation(Technomage.MOD_ID, "test2"), new ItemStackGuiIcon(Items.DIAMOND));
        view.setBackgroundTexture(new ResourceLocation(Technomage.MOD_ID, "textures/gui/technonomicon/background_02.png"));
        this.viewList.add(view);

        view = new ResearchView(new ResourceLocation(Technomage.MOD_ID, "test3"), new ItemStackGuiIcon(Items.REDSTONE));
        view.setBackgroundTexture(new ResourceLocation(Technomage.MOD_ID, "textures/gui/technonomicon/background_03.png"));
        this.viewList.add(view);

        view = new ResearchView(new ResourceLocation(Technomage.MOD_ID, "test4"), new ItemStackGuiIcon(Blocks.CRAFTING_TABLE));
        view.setBackgroundTexture(new ResourceLocation(Technomage.MOD_ID, "textures/gui/technonomicon/background_02.png"));
        this.viewList.add(view);
    }

    @Override
    public void setWorldAndResolution(Minecraft mc, int width, int height)
    {
        super.setWorldAndResolution(mc, width, height);
        this.winScaleFactor = new ScaledResolution(mc).getScaleFactor();

        this.viewList.forEach(v -> v.setBounds(new Rectangle(this.guiLeft, this.guiTop, this.guiWidth, this.guiHeight)));
        this.tabList.setPosition(this.guiLeft, this.guiTop);
        this.tabList.updateTabList(this.viewList);
    }

    @Override
    protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException
    {
        super.mouseClicked(mouseX, mouseY, mouseButton);
        if(this.tabList.getBounds().contains(mouseX, mouseY))
            this.tabList.onMouseClicked(mouseX, mouseY, mouseButton);
        else if(this.tabList.getActiveView().getBounds().contains(mouseX, mouseY))
            this.tabList.getActiveView().onMouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    protected void mouseReleased(int mouseX, int mouseY, int state)
    {
        super.mouseReleased(mouseX, mouseY, state);
        if(this.tabList.getBounds().contains(mouseX, mouseY))
            this.tabList.onMouseReleased(mouseX, mouseY, state);
        else if(this.tabList.getActiveView().getBounds().contains(mouseX, mouseY))
            this.tabList.getActiveView().onMouseReleased(mouseX, mouseY, state);
    }

    @Override
    protected void mouseClickMove(int mouseX, int mouseY, int clickedMouseButton, long timeSinceLastClick)
    {
        super.mouseClickMove(mouseX, mouseY, clickedMouseButton, timeSinceLastClick);
        if(this.tabList.getBounds().contains(mouseX, mouseY))
            this.tabList.onMouseDragged(mouseX, mouseY, clickedMouseButton, timeSinceLastClick);
        else if(this.tabList.getActiveView().getBounds().contains(mouseX, mouseY))
            this.tabList.getActiveView().onMouseDragged(mouseX, mouseY, clickedMouseButton, timeSinceLastClick);

    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks)
    {
        super.drawScreen(mouseX, mouseY, partialTicks);

        this.drawDefaultBackground();
        GlStateManager.enableBlend();

        GlStateManager.color(1.0f, 1.0f, 1.0f, 1.0f);

        // setup glScissor to only render the visible part of the view
        GL11.glEnable(GL11.GL_SCISSOR_TEST);
        GL11.glScissor((this.guiLeft + INDENT) * this.winScaleFactor, (this.guiTop + INDENT) * this.winScaleFactor,
                (this.guiWidth - INDENT * 2) * this.winScaleFactor, (this.guiHeight - INDENT * 2) * this.winScaleFactor);
        IResearchView activeView = this.tabList.getActiveView();
        if(activeView != null)
        {
            activeView.renderComponent(mouseX, mouseY);
        }
        else
        {
            Gui.drawRect(this.guiLeft, this.guiTop, this.guiWidth, this.guiHeight, 0);
            GlStateManager.color(1f, 1f, 1f);
        }
        GL11.glDisable(GL11.GL_SCISSOR_TEST);


        this.drawBorder();
        this.tabList.renderComponent(mouseX, mouseY);
        if(activeView != null && activeView.getBounds().contains(mouseX, mouseY))
            activeView.renderTooltip(mouseX, mouseY);

    }

    private void drawBorder()
    {
        int left = (this.width - this.guiWidth) / 2;
        int top = (this.height - this.guiHeight) / 2;
        this.mc.getTextureManager().bindTexture(RESEARCH_TEX);

        // we need to substract 16 from the x and y positions because of the shadow effect on the border texture
        int hCount = 1 + (this.guiWidth - 16) / 64;
        for (int i = 0; i < hCount; i++)
        {
            int rW = Math.min(64, this.guiWidth - 32 - i * 64);
            this.drawTexturedModalRect(left + 16 + i * 64, top, 48, 16, rW, 16);
            this.drawTexturedModalRect(left + 16 + i * 64, top + this.guiHeight - 16, 48, 16, rW, 16);
        }

        int vCount = 1 + (this.guiHeight - 16) / 64;
        for (int i = 0; i < vCount; i++)
        {
            int rH = Math.min(64, this.guiHeight - 32 - i * 64);
            this.drawTexturedModalRect(left, top + 16 + i * 64, 16, 48, 16, rH);
            this.drawTexturedModalRect(left + this.guiWidth - 16, top + 16 + i * 64, 16, 48, 16, rH);
        }

        // TOP LEFT
        this.drawTexturedModalRect(left + -16, top + -16, 0, 0, 48, 48);
        // TOP RIGHT
        this.drawTexturedModalRect(left + this.guiWidth - 32, top + -16, 0, 0, 48, 48);
        // BOTTOM LEFT
        this.drawTexturedModalRect(left + -16, top + this.guiHeight - 32, 0, 0, 48, 48);
        // BOTTOM RIGHT
        this.drawTexturedModalRect(left + this.guiWidth - 32, top + this.guiHeight - 32, 0, 0, 48, 48);
    }
}

File: src/main/kotlin/net/xalcon/technomage/client/gui/technonomicon/GuiTechnonomiconResearchViewer.kt
package net.xalcon.technomage.client.gui.technonomicon

import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiScreen
import net.minecraft.client.gui.ScaledResolution
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.init.Items
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.Vec2f
import net.xalcon.technomage.Technomage
import net.xalcon.technomage.api.technonomicon.IResearchPage
import net.xalcon.technomage.api.technonomicon.ResearchItem
import net.xalcon.technomage.api.technonomicon.ResearchPage
import org.lwjgl.input.Mouse
import org.lwjgl.opengl.GL11
import org.lwjgl.util.Point
import org.lwjgl.util.Rectangle

class GuiTechnonomiconResearchViewer : GuiScreen()
{
    private val RESEARCH_TEX = ResourceLocation(Technomage.MOD_ID, "textures/gui/technonomicon/research.png")

    private var winScaleFactor: Int = 0
    private var zoom = 1f
    private var x: Int = 0
    private var y: Int = 0
    private var scrolling: Boolean = false
    private var lastMouseX: Int = 0
    private var lastMouseY: Int = 0

    private var guiWidth: Int = 0
    private var guiHeight: Int = 0

    private val pageList:ArrayList<IResearchPage> = ArrayList()

    override fun doesGuiPauseGame(): Boolean = false

    override fun initGui()
    {
        super.initGui()
        this.guiWidth = 256 + 96
        this.guiHeight = 224
        val page = ResearchPage(Technomage.MOD_ID, "main", ItemStack(Items.DIAMOND_SWORD))
        this.pageList.add(page)

        page.addResearchItem(ResearchItem(Point(32, 32), "Test 01", ItemStack(Items.DIAMOND_SWORD)))
        page.addResearchItem(ResearchItem(Point(64, 64), "Test 02", ItemStack(Items.DIAMOND_SHOVEL)))
    }

    override fun setWorldAndResolution(mc: Minecraft, width: Int, height: Int)
    {
        super.setWorldAndResolution(mc, width, height)
        this.winScaleFactor = ScaledResolution(mc).scaleFactor
    }

    /**
     * Draws the screen and all the components in it.
     */
    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float)
    {
        // calculate the top + left position to center the gui
        val left = (this.width - this.guiWidth) / 2
        val top = (this.height - this.guiHeight) / 2

        // relative mouse position to the window
        val relMouseX = mouseX - left
        val relMouseY = mouseY - top

        // handle input like dragging the view
        this.handleInput(relMouseX, relMouseY)
        this.drawDefaultBackground()

        GlStateManager.pushMatrix()
        GlStateManager.enableBlend()
        GlStateManager.color(1.0f, 1.0f, 1.0f, 1.0f)
        // translate the state so 0,0 is the top left corner of our gui
        GlStateManager.translate(left.toFloat(), top.toFloat(), 0f)
        GlStateManager.pushMatrix()
        // apply the zoom, then render the view
        val INSET = 8
        val viewWidth = this.guiWidth
        val viewHeight = this.guiHeight
        val viewport = Rectangle(this.x, this.y, (viewWidth * this.zoom).toInt(), (viewHeight * this.zoom).toInt())
        val relMousePos = Vec2f(relMouseX * this.zoom, relMouseY * this.zoom)
        GlStateManager.scale(1.0f / this.zoom, 1.0f / this.zoom, 1.0f)

        GL11.glEnable(GL11.GL_SCISSOR_TEST)
        GL11.glScissor((left + INSET) * this.winScaleFactor, (top + INSET) * this.winScaleFactor, (viewWidth - INSET * 2) * this.winScaleFactor, (viewHeight - INSET * 2) * this.winScaleFactor)
        GlStateManager.color(1.0f, 1.0f, 1.0f, 1.0f)
        GlStateManager.translate(this.x.toDouble(), this.y.toDouble(), 0.0)
        this.pageList[0].renderView(viewport, relMousePos)
        GL11.glDisable(GL11.GL_SCISSOR_TEST)
        GlStateManager.popMatrix()

        GlStateManager.disableDepth()
        GlStateManager.enableBlend()
        //GlStateManager.color(1.0f, 1.0f, 1.0f, 0.3f)
        this.drawBorder()
        //GlStateManager.color(1.0f, 1.0f, 1.0f, 1.0f)
        GlStateManager.enableDepth()
        GlStateManager.popMatrix()
        super.drawScreen(mouseX, mouseY, partialTicks)
    }

    private fun handleInput(relMouseX:Int, relMouseY:Int)
    {
        if (Mouse.isButtonDown(0))
        {
            if (relMouseX > 0 && relMouseX < this.guiWidth && relMouseY > 0 && relMouseY < this.guiHeight)
            {
                if (!this.scrolling)
                {
                    // we werent scrolling before, so start scrolling mode
                    this.scrolling = true
                }
                else
                {
                    // we are scrolling, move the viewport by the relative mouse movement since last tick
                    this.x += ((relMouseX - this.lastMouseX) * this.zoom).toInt()
                    this.y += ((relMouseY - this.lastMouseY) * this.zoom).toInt()
                }

                this.lastMouseX = relMouseX
                this.lastMouseY = relMouseY
            }
        }
        else
        {
            // no longer scrolling
            this.scrolling = false
        }

        val wheel = Mouse.getDWheel()
        if (wheel < 0)
            this.zoom += 0.25f
        else if (wheel > 0)
            this.zoom -= 0.25f
        this.zoom = MathHelper.clamp(this.zoom, 1.0f, 2.0f)
    }

    private fun drawBorder()
    {
        this.mc.textureManager.bindTexture(RESEARCH_TEX)

        // we need to substract 16 from the x and y positions because of the shadow effect on the border texture
        val hCount = 1 + (this.guiWidth - 16) / 64
        for (i in 0..hCount - 1)
        {
            val rW = Math.min(64, this.guiWidth - 32 - i * 64)
            this.drawTexturedModalRect(16 + i * 64, 0, 48, 16, rW, 16)
            this.drawTexturedModalRect(16 + i * 64, this.guiHeight - 16, 48, 16, rW, 16)
        }

        val vCount = 1 + (this.guiHeight - 16) / 64
        for (i in 0..vCount - 1)
        {
            val rH = Math.min(64, this.guiHeight - 32 - i * 64)
            this.drawTexturedModalRect(0, 16 + i * 64, 16, 48, 16, rH)
            this.drawTexturedModalRect(this.guiWidth - 16, 16 + i * 64, 16, 48, 16, rH)
        }

        // TOP LEFT
        this.drawTexturedModalRect(-16, -16, 0, 0, 48, 48)
        // TOP RIGHT
        this.drawTexturedModalRect(this.guiWidth - 32, -16, 0, 0, 48, 48)
        // BOTTOM LEFT
        this.drawTexturedModalRect(-16, this.guiHeight - 32, 0, 0, 48, 48)
        // BOTTOM RIGHT
        this.drawTexturedModalRect(this.guiWidth - 32, this.guiHeight - 32, 0, 0, 48, 48)
    }
}